### PR TITLE
Allow definition of feature rendering order

### PR DESCRIPTION
**Pull request by m-kuhn**
screenshot from 2015-12-21 10-55-07

### CODE REVIEW

File: src/ui/qgsrendererv2propsdialogbase.ui [hunk 1]
```diff
@@ -133,15 +143,30 @@
...
-          <height>16</height>
-         </size>
-        </property>
-       </widget>
+      <item row="3" column="0" colspan="4">
+       <layout class="QHBoxLayout" name="horizontalLayout_3">
+        <item>
+         <widget class="QCheckBox" name="cboEnableOrderBy">
+          <property name="text">
+           <string>Control Feature rendering order</string>
```
Comment (nyalldawson) [inline src/ui/qgsrendererv2propsdialogbase.ui]: "Feature" -> "feature"

**Comment by nyalldawson** [comment]
Great feature!!

**Comment by nyalldawson** [comment]
Also - any plans on a unit test for this?

**Comment by m-kuhn** [comment]
Sure, as soon as the one for the combination orderby / setsubsetofattributes is done

**Comment by m-kuhn** [comment]
@nyalldawson all but the gui enhancement implemented

File: src/providers/virtual/qgsvirtuallayersqlitemodule.cpp [hunk 2]
```diff
@@ -588,6 +589,9 @@ int vtable_column( sqlite3_vtab_cursor *cursor, sqlite3_context* ctxt, int idx )
       case QVariant::UInt:
         sqlite3_result_int( ctxt, v.toInt() );
         break;
+      case QVariant::LongLong:
+        sqlite3_result_int64( ctxt, v.toLongLong() );
```
Comment (m-kuhn) [inline src/providers/virtual/qgsvirtuallayersqlitemodule.cpp]: @mhugo works for me this way with gdal 2

**Comment by m-kuhn** [comment]
GUI changes implemented as well, feels better to use now.
@nyalldawson I didn't move the delegate, it was easier to stick to QTableWidget instead of QTableView.

**Comment by nyalldawson** [comment]
@m-kuhn just testing now. Some feedback:

- columns in sort order dialog need headings
- I get a crash, with stack:

```
0	QVector<QVariant>::at	qvector.h	351	0x7ffff518cf4b	
1	QgsExpressionSorter::operator()	qgsfeatureiterator.cpp	38	0x7ffff53b30e2	
2	__gnu_cxx::__ops::_Val_comp_iter<QgsExpressionSorter>::operator()<QgsIndexedFeature, QList<QgsIndexedFeature>::iterator>	predefined_ops.h	166	0x7ffff53b4b48	
3	std::__unguarded_linear_insert<QList<QgsIndexedFeature>::iterator, __gnu_cxx::__ops::_Val_comp_iter<QgsExpressionSorter>>	stl_algo.h	1827	0x7ffff53b4748	
4	std::__unguarded_insertion_sort<QList<QgsIndexedFeature>::iterator, __gnu_cxx::__ops::_Iter_comp_iter<QgsExpressionSorter>>	stl_algo.h	1866	0x7ffff53b45b4	
5	std::__final_insertion_sort<QList<QgsIndexedFeature>::iterator, __gnu_cxx::__ops::_Iter_comp_iter<QgsExpressionSorter>>	stl_algo.h	1885	0x7ffff53b1e00	
6	std::__sort<QList<QgsIndexedFeature>::iterator, __gnu_cxx::__ops::_Iter_comp_iter<QgsExpressionSorter>>	stl_algo.h	1970	0x7ffff53b18ac	
7	std::sort<QList<QgsIndexedFeature>::iterator, QgsExpressionSorter>	stl_algo.h	4716	0x7ffff53b0d6b	
8	QgsAbstractFeatureIterator::setupOrderBy	qgsfeatureiterator.cpp	289	0x7ffff53b0201	
9	QgsAbstractFeatureIterator::ref	qgsfeatureiterator.cpp	223	0x7ffff53afd41	
10	QgsFeatureIterator::QgsFeatureIterator	qgsfeatureiterator.h	221	0x7fff61df2a5f	
11	QgsOgrFeatureSource::getFeatures	qgsogrfeatureiterator.cpp	414	0x7fff61e0309e	
12	QgsVectorLayerFeatureIterator::QgsVectorLayerFeatureIterator	qgsvectorlayerfeatureiterator.cpp	167	0x7ffff5575417	
13	QgsVectorLayerFeatureSource::getFeatures	qgsvectorlayerfeatureiterator.cpp	85	0x7ffff55746a9	
14	QgsVectorLayerRenderer::render	qgsvectorlayerrenderer.cpp	247	0x7ffff558b260	
15	QgsMapRendererParallelJob::renderLayerStatic	qgsmaprendererparalleljob.cpp	234	0x7ffff5436771	
16	QtConcurrent::FunctionWrapper1<void, LayerRenderJob&>::operator()	qtconcurrentfunctionwrappers.h	86	0x7ffff5438add	
17	QtConcurrent::MapKernel<QList<LayerRenderJob>::iterator, QtConcurrent::FunctionWrapper1<void, LayerRenderJob&>>::runIteration	qtconcurrentmapkernel.h	73	0x7ffff5438243	
18	QtConcurrent::MapKernel<QList<LayerRenderJob>::iterator, QtConcurrent::FunctionWrapper1<void, LayerRenderJob&>>::runIterations	qtconcurrentmapkernel.h	82	0x7ffff54382d9	
19	QtConcurrent::IterateKernel<QList<LayerRenderJob>::iterator, void>::forThreadFunction	qtconcurrentiteratekernel.h	263	0x7ffff543871a	
20	QtConcurrent::IterateKernel<QList<LayerRenderJob>::iterator, void>::threadFunction	qtconcurrentiteratekernel.h	225	0x7ffff5438087	
```

**Comment by nyalldawson** [comment]
Changing the line to `qSort( mCachedFeatures.begin(), mCachedFeatures.end(), QgsExpressionSorter( preparedOrderBys ) );` fixes it

**Comment by nyalldawson** [comment]
@m-kuhn sorry to be picky - but one last thing. Do you think the check boxes make sense for sorry order and null behaviour? I think combo boxes would be more natural and easier to instantly understand.

**Comment by m-kuhn** [comment]
@nyalldawson implemented as proposed.

For the table headers it took me quite a bit to realize that `tableWidget->clear();` kills the headers defined in the .ui :P

**Comment by nirvn** [comment]
@m-kuhn, great job!

@anitagraser , @nyalldawson , this is _very_ useful for TimeManager. I remember having point datasets that wouldn't animate that well because newer points in time would show below older ones.

**Comment by nirvn** [comment]
@m-kuhn , spotted an issue whereas the feature ordering settings aren't saved, and therefore lost on project re-load. http://hub.qgis.org/issues/14043

**Comment by SebDieBln** [comment]
I spotted two issues:
* [#14323 Checkbox _Control feature rendering order_ is not honored](http://hub.qgis.org/issues/14323)
* [#14325 Order-By-Clause accumulates default sub-clauses](http://hub.qgis.org/issues/14325)

**Pull request merged**